MINOR: Fix one flaky test in MetricsTest and improve checks for another

ijuma opened the pull request:
* Fix flakiness of `testBrokerTopicMetricsUnregisteredAfterDeletingTopic` by not
consuming messages. Filed KAFKA-5238 to track the issue that metrics for a deleted
topic may be re-created if there are fetch requests in the purgatory.

* Check the log size in `testBrokerTopicMetricsBytesInOut` before attempting to read
the `replicationBytesIn` metric. This helps understand where things have gone wrong
if if the metric has not increased (i.e. if it was an issue replicating or with the metric).

* Only remove the replication bytes in/out if the metrics are defined. This should not
affect the behaviour due to the tags, but it makes the code clearer. We've seen some
cases in Jenkins when the metric does not exist and it's still unclear how that can
happen.

ijuma committed:
MINOR: More informative assertion for flaky metrics test

mimaison commented:
Thanks for looking into this !
I've managed to reproduce the failure. The metrics are being re-created by `ReplicaManager` after the topic has been deleted:
```
...
creating kafka.producer:type=ProducerTopicMetrics,name=MessagesPerSec,topic=test-broker-topic-metric
creating kafka.producer:type=ProducerTopicMetrics,name=BytesPerSec,topic=test-broker-topic-metric
creating kafka.producer:type=ProducerTopicMetrics,name=DroppedMessagesPerSec,topic=test-broker-topic-metric
creating kafka.server:type=BrokerTopicMetrics,name=BytesRejectedPerSec,topic=test-broker-topic-metric
creating kafka.server:type=BrokerTopicMetrics,name=FailedProduceRequestsPerSec,topic=test-broker-topic-metric
creating kafka.server:type=BrokerTopicMetrics,name=FailedFetchRequestsPerSec,topic=test-broker-topic-metric
creating kafka.server:type=BrokerTopicMetrics,name=TotalProduceRequestsPerSec,topic=test-broker-topic-metric
creating kafka.server:type=BrokerTopicMetrics,name=TotalFetchRequestsPerSec,topic=test-broker-topic-metric
...
Entering topicMetricGroups
Exiting topicMetricGroups with Set(kafka.server.BrokerTopicMetrics.topic.test-broker-topic-metric)
Calling AdminUtils.deleteTopic
Calling TestUtils.verifyTopicDeletion
...
Entering in KafkaRequestHandler.close
Tags: Map(topic -> test-broker-topic-metric)
removing kafka.server:type=BrokerTopicMetrics,name=MessagesInPerSec,topic=test-broker-topic-metric
removing kafka.server:type=BrokerTopicMetrics,name=BytesInPerSec,topic=test-broker-topic-metric
removing kafka.server:type=BrokerTopicMetrics,name=BytesOutPerSec,topic=test-broker-topic-metric
removing kafka.server:type=BrokerTopicMetrics,name=BytesRejectedPerSec,topic=test-broker-topic-metric
removing kafka.server:type=BrokerTopicMetrics,name=ReplicationBytesInPerSec,topic=test-broker-topic-metric
removing kafka.server:type=BrokerTopicMetrics,name=ReplicationBytesOutPerSec,topic=test-broker-topic-metric
removing kafka.server:type=BrokerTopicMetrics,name=FailedProduceRequestsPerSec,topic=test-broker-topic-metric
removing kafka.server:type=BrokerTopicMetrics,name=FailedFetchRequestsPerSec,topic=test-broker-topic-metric
removing kafka.server:type=BrokerTopicMetrics,name=TotalProduceRequestsPerSec,topic=test-broker-topic-metric
removing kafka.server:type=BrokerTopicMetrics,name=TotalFetchRequestsPerSec,topic=test-broker-topic-metric
Exiting KafkaRequestHandler.close
creating kafka.server:type=BrokerTopicMetrics,name=MessagesInPerSec,topic=test-broker-topic-metric
java.lang.Error
	at kafka.metrics.KafkaMetricsGroup$class.newMeter(KafkaMetricsGroup.scala:86)
	at kafka.server.BrokerTopicMetrics.newMeter(KafkaRequestHandler.scala:108)
	at kafka.server.BrokerTopicMetrics.<init>(KafkaRequestHandler.scala:114)
	at kafka.server.BrokerTopicStats$$anonfun$2.apply(KafkaRequestHandler.scala:158)
	at kafka.server.BrokerTopicStats$$anonfun$2.apply(KafkaRequestHandler.scala:158)
	at kafka.utils.Pool$$anonfun$getAndMaybePut$1.apply(Pool.scala:52)
	at kafka.utils.Pool.getAndMaybePut(Pool.scala:70)
	at kafka.utils.Pool.getAndMaybePut(Pool.scala:52)
	at kafka.server.BrokerTopicStats$.getBrokerTopicStats(KafkaRequestHandler.scala:165)
	at kafka.server.ReplicaManager.kafka$server$ReplicaManager$$read$1(ReplicaManager.scala:651)
	at kafka.server.ReplicaManager$$anonfun$readFromLocalLog$1.apply(ReplicaManager.scala:746)
	at kafka.server.ReplicaManager$$anonfun$readFromLocalLog$1.apply(ReplicaManager.scala:745)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at kafka.server.ReplicaManager.readFromLocalLog(ReplicaManager.scala:745)
	at kafka.server.DelayedFetch.onComplete(DelayedFetch.scala:148)
	at kafka.server.DelayedOperation.forceComplete(DelayedOperation.scala:66)
	at kafka.server.DelayedOperation.run(DelayedOperation.scala:112)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
creating kafka.server:type=BrokerTopicMetrics,name=BytesInPerSec,topic=test-broker-topic-metric
creating kafka.server:type=BrokerTopicMetrics,name=BytesOutPerSec,topic=test-broker-topic-metric
creating kafka.server:type=BrokerTopicMetrics,name=BytesRejectedPerSec,topic=test-broker-topic-metric
creating kafka.server:type=BrokerTopicMetrics,name=FailedProduceRequestsPerSec,topic=test-broker-topic-metric
creating kafka.server:type=BrokerTopicMetrics,name=FailedFetchRequestsPerSec,topic=test-broker-topic-metric
creating kafka.server:type=BrokerTopicMetrics,name=TotalProduceRequestsPerSec,topic=test-broker-topic-metric
creating kafka.server:type=BrokerTopicMetrics,name=TotalFetchRequestsPerSec,topic=test-broker-topic-metric
creating kafka.server:type=DelayedFetchMetrics,name=ExpiresPerSec,fetcherType=follower
creating kafka.server:type=DelayedFetchMetrics,name=ExpiresPerSec,fetcherType=consumer
Entering topicMetricGroups
Exiting topicMetricGroups with Set(kafka.server.BrokerTopicMetrics.topic.test-broker-topic-metric)

```

mimaison commented:
Also if I replace the call to the deprecated `createAndShutdownStep` by `TestUtils.produceMessages`, I'm not able to hit the failure

ijuma committed:
Use `produceMessages` instead of `createAndShutdownStep`

- Use different topic names for each test
- Check log sizes in `testBrokerTopicMetricsBytesInOut` (flaky test) to narrow down causes

ijuma commented:
Thanks @mimaison. Looks like a fetch request causes the metrics to be recreated. Your suggestion to use the new producer/consumer is a good one so I made that change. I also added some additional checks to the other flaky test in that class: `testBrokerTopicMetricsBytesInOut` (suggested by @junrao).

ijuma commented:
The failure was unrelated. Also, the tests passed in Confluent Jenkins where this issue seemed to happen more frequently:

https://jenkins.confluent.io/job/kafka-branch-builder/118/

Will try again to be sure.

ijuma commented:
retest this please

ijuma commented:
Test results look good. @junrao @rajinisivaram @hachikuji, can one of you review this please?

ijuma commented:
Investigated more and found the real issue. Filed KAFKA-5238, updated the PR description and the test.

rajinisivaram commented on core/src/test/scala/unit/kafka/metrics/MetricsTest.scala:
```diff
@@ -121,6 +125,17 @@ class MetricsTest extends KafkaServerTestHarness with Logging {
+      assertTrue(s"Expected broker $brokerId to have a Log for $topicPartition with positive size, actual: $logSize",
+        logSize.map(_ > 0).getOrElse(false))
+    }
+
```
@ijuma I may be missing something, but isn't `replicationBytesIn` referring to a broker-level metric rather than a topic metric? Looking at the failure https://builds.apache.org/job/kafka-trunk-jdk7/2190/ for instance, the error says `Unable to find metric ReplicationBytesInPerSec`. This metric should have been created when the broker started up?

ijuma replied:
That's a good point @rajinisivaram and makes the failure even more surprising. Having said that, the check here is still useful in case the replication bytes in meter doesn't update.

I would like to merge this PR ASAP as `testBrokerTopicMetricsUnregisteredAfterDeletingTopic` is failing quite often in Jenkins.

ijuma replied:
I deleted my previous code as it was inaccurate. What I meant to say was, we call `removeMetric` for replication bytes in/out even if the metrics are not defined because a topic is defined (@junrao spotted this over the weekend). This should be OK because the tags would be different, but I also changed it to make it more explicit.

rajinisivaram commented on core/src/test/scala/unit/kafka/metrics/MetricsTest.scala:
```diff
@@ -74,19 +75,21 @@ class MetricsTest extends KafkaServerTestHarness with Logging {
-    createAndShutdownStep(topic, "group0", "consumer0", "producer0")
-    assertTrue("Topic metrics don't exist", checkTopicMetricsExists(topic))
+    // Produce a few messages to create the metrics
+    // Don't consume messages as it may cause metrics to be re-created causing the test to fail, see KAFKA-5238
```
After the topic is deleted, I imagine the value being added to the metrics is zero in the consume case. So it may be simple to check for zero before updating metrics. But that can be done in KAFKA-5238.

ijuma replied:
I didn't check, but that would make sense. Another option is to check if the topic exists in `ReplicaManager.allPartitions`. Indeed, let's do it in KAFKA-5328.

rajinisivaram reviewed (approved):
@ijuma Left  a comment, but I agree that it makes sense to merge this asap since it is causing build failures.

ijuma commented on core/src/test/scala/unit/kafka/metrics/MetricsTest.scala:
```diff
@@ -121,6 +125,17 @@ class MetricsTest extends KafkaServerTestHarness with Logging {
+      assertTrue(s"Expected broker $brokerId to have a Log for $topicPartition with positive size, actual: $logSize",
+        logSize.map(_ > 0).getOrElse(false))
+    }
+
```
Hmm, the bug is that we are removing `ReplicationBytesInPerSec` when we `close()`. @junrao actually spotted this bug over the weekend, but I mistakenly thought the code was doing something slightly different.

ijuma committed:
Only call `removeMetric` for replication bytes in/out if they are defined

This should not make a difference as the tags would not match, but makes
the code clearer.

ijuma commented:
The failure is unrelated:

```text
java.lang.AssertionError: expected:<List(first, second, third)> but was:<List()>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at kafka.integration.UncleanLeaderElectionTest.verifyUncleanLeaderElectionDisabled(UncleanLeaderElectionTest.scala:259)
	at kafka.integration.UncleanLeaderElectionTest.testUncleanLeaderElectionDisabled(UncleanLeaderElectionTest.scala:125)
```

Merging to trunk.
